xpublish_wms/getmap.py

```diff
@@ -0,0 +1,333 @@
+        self.bbox = [float(x) for x in query['bbox'].split(',')]
+        self.width = int(query['width'])
+        self.height = int(query['height'])
+
+        # Output style
+        self.style = query.get('styles', self.DEFAULT_STYLE)
+        self.colorscalerange = [float(x) for x in query.get('colorscalerange', 'nan,nan').split(',')]
+        self.autoscale = query.get('autoscale', "true") == "true"
+
+    def select_layer(self, ds: xr.Dataset) -> xr.DataArray:
+        """
+        Select Dataset variable, according to WMS layers request
+        :param ds:
+        :return:
+        """
+        return ds[self.parameter]
+
+    def select_time(self, da: xr.DataArray) -> xr.DataArray:
+        """
+        Ensure time selection
+
+        If time is provided :
+            - use cf_xarray to access time
+            - by default use TIME_CF_NAME
+            - method nearest to ensure at least one result
+
+        Otherwise:
+            - Get latest one
+
+        :param da:
+        :return:
+        """
+        if self.time is not None:
+            da = da.cf.sel(
+                {self.TIME_CF_NAME: self.time},
+                method="nearest"
+            )
+        else:
+            da = da.cf.isel({
+                self.TIME_CF_NAME: -1
+            })
+
+        return da
+
+    def select_custom_dim(self, da: xr.DataArray) -> xr.DataArray:
+        """
+        Select other dimension, ensuring a 2D array
+        :param da:
+        :return:
+        """
+        # Squeeze single value dimensions
+        da = da.squeeze()
+
+        # TODO: Filter dimension from custom query, if any
+
+        # Squeeze multiple values dimensions, by selecting the last value
+        for coord_name in da.cf.coords:
+            if coord_name in ("latitude", "longitude"):
+                continue
+            coord = da.cf.coords[coord_name]
+            if coord.size > 1:
+                da = da.cf.isel({coord_name: -1})
+
+        return da
+
+    def select_grid(self, da: xr.DataArray) -> xr.DataArray:
+        """
+        Select grid and reproject if needed
+        :param da:
+        :return:
+        """
+        # Some basic check for dataset
+        if not da.rio.crs:
+            da = da.rio.write_crs(self.DEFAULT_CRS)
+
+        if self.grid_is_regular(da):
+            da = self.select_regular_grid(da)
+        else:
+            da = self.select_irregular_grid(da)
+        return da
+
+    def grid_is_regular(self, da: xr.DataArray) -> bool:
+        return da.cf.coords['longitude'].dims[0] == da.cf.coords['longitude'].name
+
+    def select_regular_grid(self, da: xr.DataArray) -> xr.DataArray:
+        """
+        Filter regular grid according to WMS request :
+            - bbox
+            - width and height
+        :param da:
+        :return:
+        """
+        # TBC : need to invert lat/lon here ?
+        #minx, miny, maxx, maxy = self.bbox
+        miny, minx, maxy, maxx = self.bbox
```

Comment:
not clear for me, but after (lots of) tests that's the way it's working with my datasets

hope it's the good way to go ?

Reply:
OGC WMS requests are in the format `minx,miny,maxx,maxy` https://mapserver.org/es/ogc/wms_server.html?highlight=BBOX#test-with-a-getmap-request

This is how ncwms and every other wms system I have used works so I would rather not rock the boat on this

Reply:
Thanks for the ref example

However some others handle "reverting" bbox for particular CRS ? cf. https://github.com/ecmwf/skinnywms/blob/master/skinnywms/server.py#L29

Reply:
Interesting, thats the only one ive ever seen do that. I would rather stay on spec if possible

Reply:
I stayed on the spec in the latest commits